clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -885,16 +885,19 @@ bool Sema::DiagnoseUninstantiableTemplate(SourceLocation PointOfInstantiation,
 /// that the template parameter 'PrevDecl' is being shadowed by a new
 /// declaration at location Loc. Returns true to indicate that this is
 /// an error, and false otherwise.
-void Sema::DiagnoseTemplateParameterShadow(SourceLocation Loc, Decl *PrevDecl) {
+void Sema::DiagnoseTemplateParameterShadow(SourceLocation Loc, Decl *PrevDecl,
+                                           bool IssueWarning) {
```

Comment:
Can this be called `ShadowsFunctionOrVariableName` or something like that ? It might be clearer

Reply:
I don't care much about the exact name, although i agree with you my suggestion may not be great.  It can be `ShadowsNameOfFunctionOrVariableItIsAttachedTo` or whatever name you think is best describing of what it does. `IssueWarning` is a bit _too_ generic, or rather it forces us to think about warning in multiple places and especially someone calling `DiagnoseTemplateParameterShadow` should understand what the bool parameter does.

Reply:
I can add javadocs for the parameters that explain the semantics of `IssueWarning` and the motivating use-case for it. Is that a good compromise?

Reply:
@AaronBallman In VS20222 the javadocs show up even when they are attached to the definition... but I'll move them to the declaration to be on the safe side :)

While I agree with changing the name to `SupportedAsExtension`, I don't think the change to the diagnostic selection logic is quite right. When MSVC compatibility **is** enabled, template parameter shadowing is _always_ allowed (read: there are no cases where shadowing would be an error with MSVC compatibility enabled), so we should issue the MSVC specific diagnostic regardless of `SupportedAsExtension`.  `SupportedAsExtension` should be used to select between the error/extension diagnostic only when MSVC compatibility **is not** enabled. Perhaps we name it `SupportedForCompatibility` instead?

Reply:
> @AaronBallman In VS20222 the javadocs show up even when they are attached to the definition... but I'll move them to the declaration to be on the safe side :)

For me, that only happens if the definition is in the same TU as the use. When I tried outside of SemaTemplate.cpp (such as the uses in SemaDecl.cpp), no docs appeared in the tooltip.

> While I mostly agree with changing the name to SupportedAsExtension, I don't think the change to the diagnostic selection logic is quite right. When MSVC compatibility is enabled, template parameter shadowing is always allowed (read: there are no cases where shadowing would be an error with MSVC compatibility enabled), so we should issue the MSVC specific diagnostic regardless of SupportedAsExtension.

Good point!

> Perhaps we name it SupportedForCompatibility instead?

I could be okay with that; @cor3ntin WDYT?

Reply:
I applied the suggested changes, so this should be good to go once cor3ntin responds

Reply:
Ping @cor3ntin